### PR TITLE
pcv.visualize.pseudocolor and pcv.threshold.mask_bad

### PR DESCRIPTION
**Pull request by DannieSheng**
**Describe your changes**
Add two new inputs to the pcv.visualize.pseudocolor function such that the function is capable of taking a user-defined mask indicating locations of "bad" pixels, and create a visualization with those "bad" pixels marked with desired color (default red). The documentation is updated with sample results.

The new threshold method called "mask_bad" is highly related to the pseudocolor method, so those changes are merged here.

**Type of update**
Is this a:
* New feature or feature enhancement
* Update to documentation
* Work in progress

**Additional context**
Merged the changes which were previously in another branch, which caused some conflicts. Documentation to be included.

### CODE REVIEW

**Comment by HaleySchuhl** [comment]
@DannieSheng is ready for review? Great work getting the rest of the new code covered with tests :)

**Comment by DannieSheng** [comment]
@HaleySchuhl Oh I think I found that!

File: plantcv/plantcv/threshold/threshold_methods.py [hunk 1]
```diff
@@ -724,3 +710,48 @@ def saturation(rgb_img, threshold=255, channel="any"):
...
+    float_img = image represented by an nd-array (data type: float). Most probably, it is the result of some calculation based on the original image. So the datatype is float, and it is possible to have some "bad" values, i.e. nan and/or inf
+    bad_type = definition of "bad" type, can be 'nan', 'inf' or 'native'
+    Returns:
+    mask = A mask indicating the locations of "bad" pixels
+
+    :param float_img: np.ndarray
+    :param bad_type: str
+    :return: np.ndarray
+    """
+    params.device += 1
```
Comment (nfahlgren) [inline plantcv/plantcv/threshold/threshold_methods.py:726]:
```suggestion
```

File: plantcv/plantcv/visualize/pseudocolor.py [hunk 2]
```diff
@@ -6,20 +6,23 @@
 from matplotlib import pyplot as plt
 from plantcv.plantcv import params
 from plantcv.plantcv import fatal_error
+from plantcv.plantcv.apply_mask import apply_mask
 
 
-def pseudocolor(gray_img, obj=None, mask=None, cmap=None, background="image", min_value=0, max_value=255,
-                axes=True, colorbar=True, obj_padding="auto", title=None):
+def pseudocolor(gray_img, obj=None, mask=None, bad_mask=None, cmap=None, background="image", bad_color="red",
```
Comment (nfahlgren) [inline plantcv/plantcv/visualize/pseudocolor.py:12]: @DannieSheng, not sure we had a consensus, but what do you think about moving `bad_mask` and `bad_color` to the end of the arguments (this matches the documentation actually) to make this a non-breaking update?

File: plantcv/plantcv/threshold/__init__.py [hunk 3]
```diff
@@ -6,5 +6,6 @@
 from plantcv.plantcv.threshold.threshold_methods import texture
 from plantcv.plantcv.threshold.threshold_methods import custom_range
 from plantcv.plantcv.threshold.threshold_methods import saturation
+from plantcv.plantcv.threshold.threshold_methods import mask_bad
 
-__all__ = ["binary", "gaussian", "mean", "otsu", "triangle", "texture", "custom_range", "saturation"]
+__all__ = ["binary", "gaussian", "mean", "otsu", "triangle", "texture", "custom_range", "saturation","mask_bad"]
```
Comment (nfahlgren) [inline plantcv/plantcv/threshold/__init__.py:11]:
```suggestion
__all__ = ["binary", "gaussian", "mean", "otsu", "triangle", "texture", "custom_range", "saturation", "mask_bad"]
```

File: plantcv/plantcv/threshold/threshold_methods.py [hunk 4]
```diff
@@ -451,15 +452,8 @@ def custom_range(img, lower_thresh, upper_thresh, channel='gray'):
     params.device += 1
```
Comment (nfahlgren) [inline plantcv/plantcv/threshold/threshold_methods.py:452]:
```suggestion
```

File: plantcv/plantcv/threshold/threshold_methods.py [hunk 5]
```diff
@@ -724,3 +710,48 @@ def saturation(rgb_img, threshold=255, channel="any"):
...
+    elif bad_type.lower() == 'inf' and len(idx_inf) >= 1:
+        mask[idx_inf, idy_inf] = 255
+    # "bad" exists but not the user desired bad type, return the all-zero mask
+    else:
+        mask = mask
+        print('{} does not appear in the current image.'.format(bad_type.lower()))
+
+    _debug(visual=mask, filename=os.path.join(params.debug_outdir, str(params.device) + "_bad_mask.png"))
+
+    return mask
```
Comment (nfahlgren) [inline plantcv/plantcv/threshold/threshold_methods.py:757]:
```suggestion
    return mask

```

File: tests/tests.py [hunk 6]
```diff
@@ -943,6 +944,11 @@ def test_plantcv_parallel_process_results_invalid_json():
 TEST_THERMAL_ARRAY = "thermal_img.npz"
 TEST_THERMAL_IMG_MASK = "thermal_img_mask.png"
 TEST_INPUT_THERMAL_CSV = "FLIR2600.csv"
+TEST_BAD_MASK = "bad_mask_test.pkl"
+TEST_IM_BAD_NONE = "bad_mask_none.pkl"
+TEST_IM_BAD_BOTH = "bad_mask_both.pkl"
+TEST_IM_BAD_NAN = "bad_mask_nan.pkl"
+TEST_IM_BAD_INF = "bad_mask_inf.pkl"
```
Comment (nfahlgren) [inline tests/tests.py:951]: Couldn't we just have the masks in the data folder instead of pickle files?

File: docs/mask_bad_threshold.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
+## Mask bad pixels Threshold
+
+Creates a binary image from a grayscale based on pixel values and the definition of "bad" pixels.
```
Comment (nfahlgren) [inline docs/mask_bad_threshold.md:3]:
```suggestion
Creates a binary image from a grayscale based on pixel values and the definition of "bad" pixels.
"Bad" pixels are invalid numeric data such as not a number (nan) or infinite (inf).
```

File: docs/mask_bad_threshold.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+**plantcv.threshold.mask_bad(*img, bad_type="native"*)**
+
+**returns** thresholded/binary image
+
+- **Parameters:**
+    - img - Grayscale image data (most likely an image that is the result of some numeric calculation, e.g. a hyperspectral index image). The datatype should be "float".
```
Comment (nfahlgren) [inline docs/mask_bad_threshold.md:10]: The function input is `float_img` rather than `img`

File: docs/mask_bad_threshold.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
+## Mask bad pixels Threshold
+
+Creates a binary image from a grayscale based on pixel values and the definition of "bad" pixels.
+
+**plantcv.threshold.mask_bad(*img, bad_type="native"*)**
```
Comment (nfahlgren) [inline docs/mask_bad_threshold.md:5]: Same here, `float_img` instead of `img`?

File: docs/mask_bad_threshold.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+from plantcv import plantcv as pcv
+# Mask all types of bad pixels out present in the original image  (nan and inf)
+bad_mask_1 = pcv.threshold.mask_bad(gray_img=gray_img, bad_type="native")
+
+# Mask pixels with nan values (if any) in the original image out
+bad_mask_2 = pcv.threshold.mask_bad(gray_img=gray_img, bad_type="nan")
+
+# Mask pixels with inf values (if any) in the original image out
+bad_mask_3 = pcv.threshold.mask_bad(gray_img=gray_img, bad_type="inf")
```
Comment (nfahlgren) [inline docs/mask_bad_threshold.md:27]: `float_img` instead of `gray_img`?

File: docs/visualize_pseudocolor.md [hunk 11]
```diff
@@ -5,22 +5,25 @@ pseudocolored image. Additionally, optional maximum and minimum values can be sp
 then the image gets saved to `pcv.params.debug_outdir`, and`pcv.params.dpi` can be set for the image that gets saved. If
 unaltered, the  matplotlib default DPI is 100 pixels per inch.
 
-**plantcv.visualize.pseudocolor**(*gray_img, obj=None, mask=None, background="image", cmap=None, min_value=0, max_value=255, axes=True, colorbar=True, obj_padding='auto'*)
+**plantcv.visualize.pseudocolor**(*gray_img, obj=None, mask=None, background="image", cmap=None, min_value=0, max_value=255, axes=True, colorbar=True, obj_padding='auto', \*, bad_mask=None, bad_color="red"*)
```
Comment (nfahlgren) [inline docs/visualize_pseudocolor.md:8]:
```suggestion
**plantcv.visualize.pseudocolor**(*gray_img, obj=None, mask=None, background="image", cmap=None, min_value=0, max_value=255, axes=True, colorbar=True, obj_padding='auto', bad_mask=None, bad_color="red"*)
```

**Review by nfahlgren — CHANGES_REQUESTED** [review]
Thanks @DannieSheng! One overall note, for any function you updated to use `_debug` you can also remove the line `params.device += 1`.

File: plantcv/plantcv/threshold/threshold_methods.py (hunk 1 above)
Comment (DannieSheng) [reply plantcv/plantcv/threshold/threshold_methods.py:726]: Thank you! You are right!

File: plantcv/plantcv/visualize/pseudocolor.py (hunk 2 above)
Comment (DannieSheng) [reply plantcv/plantcv/visualize/pseudocolor.py:12]: Sure!

File: tests/tests.py (hunk 6 above)
Comment (DannieSheng) [reply tests/tests.py:951]: Did you mean save the masks as grayscale images?

**Comment by nfahlgren** [comment]
`mask_bad_threshold.md` is missing from the `mkdocs.yml` file, which is needed to be built into the documentation system

**Review by nfahlgren — APPROVED** [review]

**Pull request merged**